### PR TITLE
[mono][debugger] Fix decode_value_compute_size

### LINKED ISSUE #8653: Abort message: '* Assertion at /__w/1/s/src/mono/mono/component/debugger-agent.c:5521, condition `nfields == 0' not met'
### Android application type

.NET Android (net7.0-android, net8.0-android, etc.)

### Affected platform version

.NET 8

### Description

When working with Maui, we encountered a problem. Sometimes, if you hover the mouse over a variable, the application on Android crashes when trying to evaluate a value with error in logcat: `condition 'nfields == 0' not met`.

https://github.com/xamarin/xamarin-android/assets/48021947/95c7a4ef-50c8-47b1-b4c8-f635a6ed0f45

### Steps to Reproduce

Unfortunately, I can't provide an example to reproduce, as it doesn't reproduce on a simple application.

As I understand it, the problem occurs on this line in the `libmono-component-debugger` library:
https://github.com/mono/mono/blob/38b0227c1ce0c53058a5d78d080923435132773a/mono/mini/debugger-agent.c#L5630C2-L5630C11

Perhaps you know how to disable assertion for an application (there is some option in the `csproj ` or you can set some variable with the `adb shell`)?

### Did you find any workaround?

-

### Relevant log output

```shell
signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
01-19 16:38:12.376 20067 20067 F DEBUG   : Abort message: '* Assertion at /__w/1/s/src/mono/mono/component/debugger-agent.c:5521, condition `nfields == 0' not met'
01-19 16:38:12.376 20067 20067 F DEBUG   :     rax 0000000000000000  rbx 00007ffd445a57c8  rcx 0000711b5dd03610  rdx 0000000000000006
01-19 16:38:12.377 20067 20067 F DEBUG   :     r8  0000000000000000  r9  0000000000000000  r10 00007ffd445a57d0  r11 0000000000000203
01-19 16:38:12.377 20067 20067 F DEBUG   :     r12 000071191c2fa136  r13 0000000000000000  r14 0000000000004e2c  r15 0000000000004e2c
01-19 16:38:12.377 20067 20067 F DEBUG   :     rdi 0000000000004e2c  rsi 0000000000004e2c
01-19 16:38:12.377 20067 20067 F DEBUG   :     rbp 0000000000000004  rsp 00007ffd445a57c0  rip 0000711b5dd03610
01-19 16:38:12.377 20067 20067 F DEBUG   : 16 total frames
01-19 16:38:12.377 20067 20067 F DEBUG   : backtrace:
01-19 16:38:12.377 20067 20067 F DEBUG   :       #00 pc 0000000000061610  /apex/com.android.runtime/lib64/bionic/libc.so (abort+192) (BuildId: fa337969c798946280caa45e2d71a2e7)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #01 pc 00000000000334ab  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #02 pc 000000000003323a  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #03 pc 00000000000333c4  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #04 pc 0000000000033406  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #05 pc 0000000000027a24  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #06 pc 0000000000019a4d  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #07 pc 000000000002796e  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #08 pc 0000000000019a4d  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #09 pc 00000000000187b2  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #10 pc 0000000000015961  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #11 pc 0000000000018005  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #12 pc 000000000001784f  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #13 pc 000000000002c6c7  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #14 pc 0000000000023685  /data/app/~~MUkR16U5S80Q4lyx5vlMKQ==/com.Maui.Controls.Tests-JXpS9RnRnPX8B8uFbokXRw==/lib/x86_64/libmono-component-debugger.so (BuildId: 1c74febae626d008490d4a981f019b38025930af)
01-19 16:38:12.377 20067 20067 F DEBUG   :       #15 pc 000000000000007e  <anonymous:41f0e000>
```

### PR DESCRIPTION
**Pull request by thaystg**
Implement decode_fixed_size_array_compute_size_internal
We were skipping this kind of field while was computing the size and this was causing a weird behavior because we were reading wrong information from the buffer.

Fix decode_value_compute_size it was decoding the buffer wrongly.

Improve wasm debugger - support inspect fixed size array (necessary to add a unit test).

Fixes https://github.com/xamarin/xamarin-android/issues/8653

### CODE REVIEW

File: src/mono/mono/component/debugger-agent.c [hunk 1]
```diff
@@ -5372,8 +5372,8 @@ decode_vtype (MonoType *t, MonoDomain *domain, gpointer void_addr, gpointer void
 	}
 
 	nfields = decode_int (buf, &buf, limit);
-	while ((f = mono_class_get_fields_internal (klass, &iter))) {
-		if (G_UNLIKELY (!f->type)) {
+		while ((f = mono_class_get_fields_internal (klass, &iter))) {
+				if (G_UNLIKELY (!f->type)) {
```
Comment (lambdageek) [inline src/mono/mono/component/debugger-agent.c:5376]: nit: extra whitespace

File: src/mono/mono/component/debugger-agent.c [hunk 2]
```diff
@@ -5487,27 +5534,27 @@ decode_vtype_compute_size (MonoType *t, MonoDomain *domain, gpointer void_buf, g
...
-		if (mono_field_is_deleted (f))
-			continue;
-		
+							continue;
+			}
+				if (f->type->attrs & FIELD_ATTRIBUTE_STATIC)
+					continue;
+				if (mono_field_is_deleted (f))
+					continue;
+
```
Comment (lambdageek) [inline src/mono/mono/component/debugger-agent.c:5549]: also here

File: src/mono/mono/component/debugger-agent.c [hunk 3]
```diff
@@ -5487,27 +5534,27 @@ decode_vtype_compute_size (MonoType *t, MonoDomain *domain, gpointer void_buf, g
...
 		gboolean cur_field_in_extra_space = from_by_ref_value_type;
   		gboolean members_in_extra_space = cur_field_in_extra_space || m_type_is_byref (f->type) || m_class_is_byreflike (mono_class_from_mono_type_internal (f->type));
 		int field_size = decode_value_compute_size (f->type, 0, domain, buf, &buf, limit, members_in_extra_space);
 		if (members_in_extra_space)
 			ret += field_size;
 		if (err != ERR_NONE)
-			return err;
-		nfields --;
+					return err;
+				nfields --;
```
Comment (lambdageek) [inline src/mono/mono/component/debugger-agent.c:5557]: and here

File: src/mono/mono/component/debugger-agent.c [hunk 4]
```diff
@@ -5530,7 +5577,7 @@ decode_value_compute_size (MonoType *t, int type, MonoDomain *domain, guint8 *bu
 {
 	if (type == 0)
 		type = decode_byte (buf, &buf, limit);
-	int ret = 0;
+		int ret = 0;
```
Comment (lambdageek) [inline src/mono/mono/component/debugger-agent.c:5580]: and here

**Review by lambdageek — APPROVED** [review]
native code lgtm, except there's some extra indentation

**Comment by lambdageek** [comment]
@ilonatommy could you take a look at the debug proxy

**Comment by lewing** [comment]
does this need to be backported?

**Comment by thaystg** [comment]
> does this need to be backported?

Yes. I will do it.

**Pull request merged**